rust/wal3/src/writer.rs

```diff
@@ -0,0 +1,353 @@
+            tokio::sync::oneshot::Sender<Result<LogPosition, Error>>,
+        )>,
+    ) {
+        let mut messages = Vec::with_capacity(work.len());
+        let mut notifies = Vec::with_capacity(work.len());
+        for work in work.into_iter() {
+            messages.push(work.0);
+            notifies.push(work.1);
+        }
+        match self
+            .append_batch_internal(fragment_seq_no, log_position, messages)
+            .await
+        {
+            Ok(log_position) => {
+                for (idx, notify) in notifies.into_iter().enumerate() {
+                    let log_position = log_position + idx;
+                    if notify.send(Ok(log_position)).is_err() {
+                        // TODO(rescrv):  Counter this.
+                    }
+                }
+            }
+            Err(e) => {
+                for notify in notifies {
+                    if notify.send(Err(e.clone())).is_err() {
+                        // TODO(rescrv):  Counter this.
+                    }
+                }
+            }
+        }
+    }
+
+    async fn append_batch_internal(
+        &self,
+        fragment_seq_no: FragmentSeqNo,
+        log_position: LogPosition,
+        messages: Vec<Vec<u8>>,
+    ) -> Result<LogPosition, Error> {
+        assert!(!messages.is_empty());
+
+        // Construct the columns.
+        let messages_len = messages.len();
+        let mut positions = Vec::with_capacity(messages_len);
+        let mut bodies = Vec::with_capacity(messages_len);
+        for (index, message) in messages.iter().enumerate() {
+            let position = log_position + index;
+            positions.push(position);
+            bodies.push(message.as_slice());
+        }
+        let offsets = positions.iter().map(|p| p.offset).collect::<Vec<_>>();
+        let timestamps_us = positions.iter().map(|p| p.timestamp_us).collect::<Vec<_>>();
+        let offsets = UInt64Array::from(offsets);
+        let timestamps_us = UInt64Array::from(timestamps_us);
+        let bodies = BinaryArray::from(bodies);
+        // SAFETY(rescrv):  The try_from_iter call will always succeed.
+        // TODO(rescrv):  Arrow pre-allocator.
+        let batch = RecordBatch::try_from_iter(vec![
+            ("offset", Arc::new(offsets) as ArrayRef),
+            ("timestamp_us", Arc::new(timestamps_us) as ArrayRef),
+            ("body", Arc::new(bodies) as ArrayRef),
+        ])
+        .unwrap();
+
+        // Write to parquet.
+        let props = WriterProperties::builder()
+            .set_compression(Compression::SNAPPY)
+            .build();
+        let mut buffer = vec![];
+        let mut writer = ArrowWriter::try_new(&mut buffer, batch.schema(), Some(props)).unwrap();
+        writer.write(&batch).map_err(Arc::new)?;
+        writer.close().map_err(Arc::new)?;
+        let setsum = Setsum::default();
+
+        // Upload the log.
+        let path = format!(
+            "{}/log/Bucket={}/FragmentSeqNo={}.parquet",
```

Comment:
nit pull this formatting into a util so that if it gets used elsewhere we have it in a central place